File: data_structures/week1_basic_data_structures/check_brackets_in_code/check_brackets.py
import sys
from typing import Final, Literal, Optional, cast

OPENING_BRACKETS: Final = ("[", "{", "(")

OpeningBracketType = Literal["[", "{", "("]


class Bracket:
    def __init__(self, opening_bracket_type: OpeningBracketType, *, pos: int) -> None:
        self.__opening_bracket_type: Final = opening_bracket_type
        self.position: Final = pos

    def match(self, closing_bracket_type: str) -> bool:  # noqa: FNE005
        return (
            (self.__opening_bracket_type == "[" and closing_bracket_type == "]")
            or (self.__opening_bracket_type == "{" and closing_bracket_type == "}")
            or (self.__opening_bracket_type == "(" and closing_bracket_type == ")")
        )


def check_brackets(text: str) -> Optional[int]:  # noqa: C901
    opening_brackets_stack: list[Bracket] = []

    for i, next_char in enumerate(text, start=1):
        if next_char in ("(", "[", "{"):
            opening_brackets_stack.append(Bracket(cast(OpeningBracketType, next_char), pos=i))

        elif next_char in (")", "]", "}"):  # noqa: SIM102
            if not opening_brackets_stack or not opening_brackets_stack.pop().match(next_char):
                return i

    if opening_brackets_stack:
        return opening_brackets_stack[0].position

    return None


if __name__ == "__main__":
    text = sys.stdin.read()

    print(check_brackets(text) or "Success")

File: data_structures/week1_basic_data_structures/check_brackets_in_code/check_brackets.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Bracket {
  char type;
  int position;
} Bracket;

bool match(char opening_bracket, char closing_bracket) {
  return (opening_bracket == '(' && closing_bracket == ')') ||
         (opening_bracket == '[' && closing_bracket == ']') ||
         (opening_bracket == '{' && closing_bracket == '}');
}

int check_brackets(char text[]) {
  Bracket opening_brackets_stack[strlen(text)];
  int stack_top = -1;

  for (int i = 0; i < strlen(text); i++) {
    char next_char = text[i];
    if (next_char == '(' || next_char == '[' || next_char == '{') {
      Bracket b;
      b.type = next_char;
      b.position = i + 1;
      opening_brackets_stack[++stack_top] = b;
    } else if (next_char == ')' || next_char == ']' || next_char == '}') {
      if (stack_top == -1)
        return i + 1;

      Bracket top = opening_brackets_stack[stack_top];
      if (match(top.type, next_char))
        stack_top--;
      else
        return i + 1;
    }
  }

  if (stack_top == -1)
    return 0;
  else
    return opening_brackets_stack[0].position;
}

int main() {
  char text[100001];
  fgets(text, 100001, stdin);
  int result = check_brackets(text);

  if (result == 0)
    printf("Success\n");
  else
    printf("%d\n", result);

  return 0;
}
